### ISSUE TITLE
`:foo/class` and `:bar/class` should not remove each other's classes

### ISSUE THREAD

**Issue by thedavidmeister** [hoplon/hoplon#202]
thedavidmeister [3:33 AM] 
@flyboarder the ??class test i linked to

flyboarder [3:44 AM] 
@thedavidmeister looks good :thumbsup:

----- Today October 1st, 2017 -----
hiskennyness [6:00 AM] 
joined #hoplon.

thedavidmeister [2:23 PM] 
@flyboarder well it fails :stuck_out_tongue:

[2:23] 
`:baz/class` overrides `:class` rather than adding to it

[2:24] 
actually i'm not aware of a good way to build up classes on an element without juggling data structures :disappointed:

[2:27] 
i often find myself wanting to do something like this

[2:27] 
 ```
(defelem my-basic-foo [_ _] (div :class "foo"))
(defelem my-edge-case-foo [_ _] (my-basic-foo :edge-case/class "edge-case"))
```

[2:27] 
and get back a `div.foo.edge-case`

flyboarder [2:57 PM] 
@thedavidmeister I'm not sure it should work that way, usually my edge cases grab the existing class before modifying it, that's specific to the edge attribute tho

[2:57] 
It makes sense that it overrides it

thedavidmeister [2:58 PM] 
that's a lot of boilerplate...

flyboarder [2:58 PM] 
As it's a namespaced version of the generic class attribute

thedavidmeister [2:58 PM] 
well no, it makes sense that `:class` overrides `:class` but i wouldn't expect `:foo/class` to override `:bar/class`

flyboarder 
[2:59 PM] 
I would depending on the implementation, :foo/class may/may not preserve the existing class

[3:00] 
Same with :bar/class

thedavidmeister [3:00 PM] 
the default implementation

[3:01] 
it seems a lot more useful to be able to merge classes than have later calls blow away earlier calls

flyboarder [3:02 PM] 
That would depend on the CSS lib tho wouldn't it? My thought is that we just can't test for these things and the attribute providers should have proper tests for their implementation. However I see the value in merging the data.

thedavidmeister [3:03 PM] 
you can test for your expectations using the implementation provided by hoplon

[3:04] 
this isn't about testing the lib

[3:04] 
this is about what hoplon does

flyboarder [3:04 PM] 
Right the default is a "set" style

thedavidmeister [3:05 PM] 
yes, but it calls that "set" once for each different `:x/class`

[3:06] 
i could imagine a hoplon implementation that tracks the classes used for each ns and has the merge behaviour in it

[3:06] 
using the same lib

[3:07] 
it already has some logic around setting `true`/`false` to add/remove classes, so it could probably extend that

flyboarder 
[3:08 PM] 
Yeah, ok so that makes sense, if I use a non hoplon element with a :class call I would expect any existing classes to remain, so we should probably be merging

thedavidmeister [3:09 PM] 
mmm basically

[3:09] 
that's what i'm thinking

[3:09] 
that it just doesn't touch any classes it doesn't know about

flyboarder 
[3:09 PM] 
Yep I'm on track with that

[3:10] 
Can you open a ticket to remind me :)

thedavidmeister [3:10 PM] 
sure thing

[3:11] 
"merging" might even be more sophisticated than what i'm suggesting :stuck_out_tongue:

[3:11] 
it's really just about not blowing away unrelated classes that already exist

**Comment by meelijane** [comment]
Yes please! :+1:

**Comment by flyboarder** [comment]
fixed in 7.1-proposal
